File: 백준/Gold/1947. 선물 전달/선물 전달.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    private static final long MOD = 1_000_000_000;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        long[] dp = new long[n + 1];
        dp[0] = 1;

        if (n >= 1) dp[1] = 0;
        if (n >= 2) dp[2] = 1;

        for (int i = 3; i <= n; i++) {
            dp[i] = ((i - 1) * (dp[i - 1] + dp[i - 2])) % MOD;
        }

        System.out.println(dp[n]);
    }
}

File: 백준/Gold/1947. 선물 전달/선물 전달.kt
import java.io.BufferedReader
import java.io.InputStreamReader

private const val MOD = 1_000_000_000L

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    val dp = LongArray(n + 1)
    dp[0] = 1

    if (n >= 1) dp[1] = 0
    if (n >= 2) dp[2] = 1

    for (i in 3..n) {
        dp[i] = ((i - 1) * (dp[i - 1] + dp[i - 2])) % MOD
    }

    println(dp[n])
}
